aep/general/0004/aep.md

```diff
@@ -0,0 +1,101 @@
+# Resource types
+
+Most APIs expose _resources_ (their primary nouns) which users are able to
+create, retrieve, and manipulate. APIs are allowed to name their resource types
+as they see fit, and are only required to ensure uniqueness within that API.
+This means that it is possible (and often desirable) for different APIs to use
+the same type name. For example, a Memcache and Redis API would both want to
+use `Instance` as a type name.
+
+When mapping the relationships between APIs and their resources, however, it
+becomes important to have a single, globally-unique type name. Additionally,
+tools such as Kubernetes or GraphQL interact with APIs from multiple providers.
+
+## Guidance
+
+APIs **must** define a resource type for each resource in the API, according to
+the following pattern: `{API Name}/{Type}`. The type name:
+
+- **must** Start with an uppercase letter.
+- **must** Only contain alphanumeric characters.
+- **must** Be of the singular form of the noun.
+- **must** Use PascalCase (UpperCamelCase).
+- For Kubernetes, the type name **must** match the [object][] name.
+- For Protobuf, the type name **must** match the name of the protobuf message.
+
+### Examples
+
+Examples of resource types include:
+
+- `pubsub.googleapis.com/Topic`
+- `pubsub.googleapis.com/Subscription`
+- `spanner.googleapis.com/Database`
+- `spanner.googleapis.com/Instance`
+- `networking.istio.io/Instance`
+
+### Annotating resource types
+
+APIs **must** annotate the resource types for each resource in the API using
+the [`google.api.resource`][resource] annotation:
+
+```proto
+// A representation of a Pub/Sub topic.
+message Topic {
+  option (google.api.resource) = {
+    type: "pubsub.googleapis.com/Topic"
+    pattern: "projects/{project}/topics/{topic}"
+    singular: "topic"
+    plural: "topics"
+  };
+
+  // Name and other fields...
+}
+```
+
+- Patterns **must** correspond to the [resource path][resource-paths].
+- Pattern variables (the segments within braces) **must** use `snake_case`, and
+  **must not** use an `_id` suffix.
+- Pattern variables **must** conform to the format `[a-z][_a-z0-9]*[a-z0-9]`.
```

Comment:
That's not invalid, I believe - In practice it maps to a protobuf field name, and I think a_____9 is valid?

Reply:
I agree that's a valid field name, but pattern variables don't correspond to a field name, they correspond to a the name of a resource type (which should be in PascalCase, and converting PascalCase to snake_case never results in adjacent underscores).

Reply:
corresponding to a resource type seems reasonable - it may also avoid the issue where it's hard to figure our the parent resource types in the pattern, since the collection names can be short-hand for the resource (that clause in the resource path about https://aep.dev/122#nested-collections)

In that case we would have stuff like:

```/users/{user}/events/{userEvent}```

is that ok with you?

Reply:
Actually I was thinking through this a little more - and I think that in general, patterns a somewhat flawed - they force you to re-enumerate the resource paths of parents, or worse have a parent resource path that doesn't match.

Really the only thing a sub-resource can do is add a suffix to a specific parent. So really, I think we should do something like:

```
google.api.resource = {
    type: "user-event"
    singular: "user-event"
    plural: "user-events"
    parents = [ 
        type: "user",
        path_suffix = "/events/{user-event}"
    ]
}
```

does that make sense? I think we'd have to come up with our own resource annotation to support it, but it allows for generation of correct, fully qualified resource paths.

Reply:
the pattern annotation is used by the api-linter, but it's not used for gRPC-HTTP bindings at all.

My opinion is that we should fork the annotation here so it's a bit more semantically clear, and work from parent references. But @rofrankel noted in our last meeting that having all patterns enumerated is nice from a documentation perspective.

We've agreed upon keeping the existing annotation as-is, but introducing a more standard resource definition syntax into the AEPs.

I might try to just fix this AEP to work without it though (e.g. documenting the pattern syntax as suggested) so that we can get this pretty critical AEP in while we debate resource syntax.

Reply:
fixed the aep as-is for now, I'll work on adding the resource definition at a future time.